pypeit/core/wavecal/kdtree_generator.py

```diff
@@ -0,0 +1,297 @@
+            for x in range(l + 1, ll - 2):
+                for xx in range(x + 1, ll - 1):
+                    index[cnt, 0] = l
+                    index[cnt, 1] = x
+                    index[cnt, 2] = xx
+                    index[cnt, 3] = ll
+                    pattern[cnt, 0] = (linelist[x] - linelist[l]) / (linelist[ll] - linelist[l])
+                    pattern[cnt, 1] = (linelist[xx] - linelist[l]) / (linelist[ll] - linelist[l])
+                    cnt += 1
+    return pattern, index
+
+
+@nb.jit(nopython=True, cache=True)
+def pentagon(linelist, numsrch, maxlin):
+    """
+    linelist : list of wavelength calibration lines (must be sorted by ascending wavelength)
+    """
+    nptn = 5  # Number of lines used to create a pattern
+
+    sz_l = linelist.shape[0]
+
+    # Count the number of patterns that will be created
+    cnt = 0
+    for l in range(0, sz_l - nptn + 1):
+        nup = (l + nptn - 1) + numsrch
+        if nup > sz_l: nup = sz_l
+        for ll in range(l + nptn - 1, nup):
+            if (linelist[ll] - linelist[l]) > maxlin: continue
+            for x in range(l + 1, ll - 3):
+                for xx in range(x + 1, ll - 2):
+                    for xxx in range(xx + 1, ll - 1):
+                        cnt += 1
+
+    index = np.zeros((cnt, nptn), dtype=nb.types.uint64)
+    pattern = np.zeros((cnt, nptn - 2), dtype=nb.types.float64)
+
+    # Generate the patterns
+    cnt = 0
+    for l in range(0, sz_l - nptn + 1):
+        nup = (l + nptn - 1) + numsrch
+        if nup > sz_l: nup = sz_l
+        for ll in range(l + nptn - 1, nup):
+            if (linelist[ll] - linelist[l]) > maxlin: continue
+            # Create a pattern with these two endpoints
+            for x in range(l + 1, ll - 3):
+                for xx in range(x + 1, ll - 2):
+                    for xxx in range(xx + 1, ll - 1):
+                        index[cnt, 0] = l
+                        index[cnt, 1] = x
+                        index[cnt, 2] = xx
+                        index[cnt, 3] = xxx
+                        index[cnt, 4] = ll
+                        pattern[cnt, 0] = (linelist[x] - linelist[l]) / (linelist[ll] - linelist[l])
+                        pattern[cnt, 1] = (linelist[xx] - linelist[l]) / (linelist[ll] - linelist[l])
+                        pattern[cnt, 2] = (linelist[xxx] - linelist[l]) / (linelist[ll] - linelist[l])
+                        cnt += 1
+    return pattern, index
+
+
+@nb.jit(nopython=True, cache=True)
+def hexagon(linelist, numsrch, maxlin):
+    """
+    linelist : list of wavelength calibration lines (must be sorted by ascending wavelength)
+    """
+
+    # Number of lines used to create a pattern
+    nptn = 6
+
+    sz_l = linelist.shape[0]
+    # Count the number of patterns that will be created
+    cnt = 0
+    for l in range(0, sz_l - nptn + 1):
+        nup = (l + nptn - 1) + numsrch
+        if nup > sz_l: nup = sz_l
+        for ll in range(l + nptn - 1, nup):
```

Comment:
I trust that these nested `for` loops are needed.
I guess either way that `nb` speeds them up just fine.

Reply:
That's right - this routine is fast because of numba.